business/workloads.go

```diff
@@ -2451,51 +2381,55 @@ func (in *WorkloadService) updateWorkload(ctx context.Context, cluster string, n
-		kubernetes.Deployments,
-		kubernetes.ReplicaSets,
-		kubernetes.ReplicationControllers,
-		kubernetes.DeploymentConfigs,
-		kubernetes.StatefulSets,
-		kubernetes.Jobs,
-		kubernetes.CronJobs,
-		kubernetes.Pods,
-		kubernetes.DaemonSets,
+	workloadGVKs := map[schema.GroupVersionKind]ctrlclient.Object{
+		kubernetes.Deployments:            &apps_v1.Deployment{},
+		kubernetes.ReplicaSets:            &apps_v1.ReplicaSet{},
+		kubernetes.ReplicationControllers: &core_v1.ReplicationController{},
+		kubernetes.DeploymentConfigs:      &osapps_v1.DeploymentConfig{},
+		kubernetes.StatefulSets:           &apps_v1.StatefulSet{},
+		kubernetes.Jobs:                   &batch_v1.Job{},
+		kubernetes.CronJobs:               &batch_v1.CronJob{},
+		kubernetes.Pods:                   &core_v1.Pod{},
+		kubernetes.DaemonSets:             &apps_v1.DaemonSet{},
 	}
 
 	// workloadGVK is an optional parameter used to optimize the workload type fetch
 	// By default workloads use only the "name" but not the pair "name,type".
 	if workloadGVK.Kind != "" {
-		found := false
-		for _, wt := range workloadGVKs {
-			if workloadGVK.Kind == wt.Kind {
-				found = true
-				break
-			}
-		}
+		v, found := workloadGVKs[workloadGVK]
 		if found {
-			workloadGVKs = []schema.GroupVersionKind{workloadGVK}
+			workloadGVKs = map[schema.GroupVersionKind]ctrlclient.Object{workloadGVK: v}
 		}
 	}
 
 	wg := sync.WaitGroup{}
 	wg.Add(len(workloadGVKs))
 	errChan := make(chan error, len(workloadGVKs))
 
-	for _, workloadGVK := range workloadGVKs {
-		go func(wkGVK schema.GroupVersionKind) {
+	// TODO: We should always pass the GVK. We should NOT be patching every kind with the same name/namespace.
+	for workloadGVK, workloadObj := range workloadGVKs {
+		go func(wkGVK schema.GroupVersionKind, obj ctrlclient.Object) {
 			defer wg.Done()
-			var err error
 			if in.isWorkloadIncluded(wkGVK.Kind) {
-				err = userClient.UpdateWorkload(namespace, workloadName, wkGVK, jsonPatch, patchType)
-			}
-			if err != nil {
-				if !errors.IsNotFound(err) {
+				if err := userClient.UpdateWorkload(namespace, workloadName, workloadObj, jsonPatch, patchType); err != nil && !errors.IsNotFound(err) {
 					log.Errorf("Error fetching %s per namespace %s and name %s: %s", wkGVK, namespace, workloadName, err)
 					errChan <- err
+					return
+				}
+
+				kubeCache, err := in.cache.GetKubeCache(cluster)
+				if err != nil {
+					log.Errorf("Unable to find kube cache for cluster: %s. You may see stale data but the update was processed correctly.", cluster)
+					return
+				}
+
+				if err := kubernetes.WaitForObjectUpdateInCache(ctx, kubeCache, obj); err != nil {
+					// It won't break anything if we return the object before it is updated in the cache.
+					// We will just show stale data so just log an error here instead of failing.
+					log.Errorf("Failed waiting for object to update in cache. You may see stale data but the update was processed correctly. Error: %s", err)
```

Comment:
Tweak msg to make it unique
```suggestion
					log.Errorf("Failed to wait for object to update in cache. You may see stale data but the update was processed correctly. Error: %s", err)
```